### PR TITLE
Setup

### PR DESCRIPTION
**Pull request by manumanuk**
# Purpose
* Create initial project setup for driver testing with an Arduino
* Write SPI interfacing modes and unit tests for each type

# New Changes
* Created project structure, including one folder for quick_tests just to check if the CC1120 can read data appropriately, and one folder for fully-fledged unit testing & SPI functions
* Created SPI read, write, burst read/write, FIFO standard access, and FIFO direct access functions as specified in datasheet
* Created unit tests for each SPI function that can be run on the CC1120 to check if they are written correctly
* Documentation + documentation standard

# Outstanding Changes
* Creating an MVP transmit/receive program
* Still need to figure out best practices in terms of unit testing & program organization

### CODE REVIEW

File: arduino_unit_tests/arduino_unit_tests.ino [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    delay(1000);
+
+    Serial.println("Starting unit tests...");
+    bool status = true;
+    status &= checkStrobe();
+    status &= checkSPIRead();
+    status &= checkSPIWrite();
+    status &= checkFIFOReadWrite();
+    if (!status)
+        Serial.println("CC1120 tests failed. Please try grounding RST and pulling it high again.");
```
Comment (sahil-kale) [inline arduino_unit_tests/arduino_unit_tests.ino]: hook it up to a gpio pin lol

File: arduino_unit_tests/cc1120_ext_regs.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
+#ifndef EXT_REGS_CC1120_H
+#define EXT_REGS_CC1120_H
+
+#define IF_MIX_CFG          0x00
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_ext_regs.h]: C style way of writing these but if you do #define make sure you make them UNSIGNED (ex: 0x00U) to avoid weird errors

File: arduino_unit_tests/cc1120_spi.cpp [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+#include <SPI.h>
+
+/**
+ * @brief Reads a register from the CC1120.
+ * 
+ * @param addr - The address of the register to read.
+ * @param data - Pointer to the variable to store the read data.
+ * @return true - If the read was successful.
+ * @return false - If the register is not valid, or the status byte is invalid.
+ */
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: why are these not in the header file lol

File: arduino_unit_tests/cc1120_spi.cpp [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+bool arduinoReadSPI(uint8_t addr, uint8_t *data) {
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Keep one API not 2

File: arduino_unit_tests/cc1120_spi.cpp [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+    return true;
+}
+
+/**
+ * @brief - Reads consecutive registers from the CC1120 in burst access mode.
+ * 
+ * @param addr - The address of the first register to read.
+ * @param data - The array to store the read data.
+ * @param len - The number of registers to read.
+ * @return true - If the read was successful.
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Improvement: make a specific return typedef so you can add more status codes in the future

File: arduino_unit_tests/cc1120_spi.cpp [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+ * @return false - If the register is not valid, or the status byte is invalid.
+ */
+bool arduinoBurstReadSPI(uint8_t addr, uint8_t data[], uint8_t len) {
+    if(addr >= EXT_ADDR) {
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: why equal to?

File: arduino_unit_tests/cc1120_spi.cpp [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+        Serial.println("Not a valid register!");
+        return false;
+    }
+
+    union cc_st ccstatus;
+    digitalWrite(CS, LOW);
+    if (!sendByteReceiveStatus(R_BIT | BURST_BIT | addr)) {
+        digitalWrite(CS, HIGH);
+        return false;
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Comment please

File: arduino_unit_tests/cc1120_spi.cpp [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+bool sendByteReceiveStatus(uint8_t data) {
+    union cc_st ccstatus;
+
+    uint8_t i=1;
+    for (; i <= 5; i++) {
+        ccstatus.v = SPI.transfer(data);
+        if (ccstatus.ccst.chip_ready == 1) {
+            Serial.print("CC1120 chip not ready. Retrying... (");
+            Serial.print(i);
+            Serial.println("/5)");
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Define debug ifdef so you don't get random serial printlns

File: arduino_unit_tests/cc1120_spi.cpp [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+    digitalWrite(CS, LOW);
+    if (!sendByteReceiveStatus(R_BIT | EXT_ADDR)) {
+        digitalWrite(CS, HIGH);
+        return false;
+    }
+    SPI.transfer(addr);
+    *data = SPI.transfer(0xFF);
+    digitalWrite(CS, HIGH);
+    return true;
+}
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Add comment sjustifying why

File: arduino_unit_tests/cc1120_spi.cpp [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+    }
+
+    digitalWrite(CS, LOW);
+    if (!sendByteReceiveStatus(addr)) {
+        digitalWrite(CS, HIGH);
+        return false;
+    }
+    if (!sendByteReceiveStatus(data)) {
+        digitalWrite(CS, HIGH);
+        return false;
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.cpp]: Remove early return

Comment (sahil-kale) [reply arduino_unit_tests/cc1120_spi.cpp]: bool status = true;

if(status)

if(status)

//Cleanup at the end

File: arduino_unit_tests/cc1120_spi.h [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+struct cc_status {
+  uint8_t res : 4;
+  uint8_t state : 3;
+  uint8_t chip_ready : 1;
+};
+
+union cc_st {
+  struct cc_status ccst;
+  uint8_t v;
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_spi.h]: ?

File: arduino_unit_tests/cc1120_unit_tests.h [hunk 12]
```diff
@@ -0,0 +1,13 @@
+#ifndef UNIT_TESTS_CC1120_H
+#define UNIT_TESTS_CC1120_H
+
+#include <stdbool.h>
+#include <stdint.h>
```
Comment (sahil-kale) [inline arduino_unit_tests/cc1120_unit_tests.h]: why is this included?

**Review by sahil-kale — CHANGES_REQUESTED** [review]

File: cc1120_arduino_tests/cc1120_spi.h [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#define R_BIT 1 << 7
+#define BURST_BIT 1 << 6
+
+struct cc_status {
+  uint8_t res : 4;
+  uint8_t state : 3;
```
Comment (dgobalak) [inline cc1120_arduino_tests/cc1120_spi.h]: It might help to have an enum with all the 8 possible states (TX, RX, Fast TX, etc.)

File: cc1120_arduino_tests/cc1120_spi.cpp [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+ * @param data - The data to write to the register.
+ * @return true - If the write was successful.
+ * @return false - If the register is not valid, or the status byte is invalid.
+ */
+bool arduinoWriteFIFODirect(uint8_t addr, uint8_t data) {
+    if (addr < CC1120_REGS_FIFO_TX_START || addr > CC1120_REGS_FIFO_RX_END) {
+        Serial.println("Not a valid FIFO register!");
+        return false;
+    }
+
```
Comment (dgobalak) [inline cc1120_arduino_tests/cc1120_spi.cpp]: The FIFO on the CC1120 holds 128 bytes max. We should check if the length of the data exceeds that.

File: cc1120_arduino_tests/cc1120_spi.h [hunk 15]
```diff
@@ -0,0 +1,203 @@
+#ifndef SPI_CC1120_H
+#define SPI_CC1120_H
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#define R_BIT 1 << 7
+#define BURST_BIT 1 << 6
+
```
Comment (dgobalak) [inline cc1120_arduino_tests/cc1120_spi.h]: Define the FIFO size (should be 128 bytes based on the datasheet) so we can compare with the length of transmitted data.

File: cc1120_arduino_tests/cc1120_spi.h [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+struct cc_status {
+  uint8_t res : 4;
+  uint8_t state : 3;
+  uint8_t chip_ready : 1;
+};
+
```
Comment (dgobalak) [inline cc1120_arduino_tests/cc1120_spi.h]: We could have an enum for the error codes too

File: cc1120_arduino_tests/cc1120_spi.h [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+union cc_st {
+  struct cc_status ccst;
+  uint8_t data;
+};
+
```
Comment (dgobalak) [inline cc1120_arduino_tests/cc1120_spi.h]: Define the part number (0x48) and part version number. It should be helpful for testing reads from the extended register space. 
Something like:
#define CC1120_PART_NUM  0x48U;
#define CC1120_PART_VER  ?; // Idk what the part version is lol

Comment (dgobalak) [reply cc1120_arduino_tests/cc1120_spi.h]: Or you can make a struct to hold this info

Comment (manumanuk) [reply cc1120_arduino_tests/cc1120_spi.h]: This is all defined in the cc1120_regs file, is that fine?

Comment (dgobalak) [reply cc1120_arduino_tests/cc1120_spi.h]: I don't see them there. The extended addresses (CC1120_REGS_EXT_PARTNUMBER and CC1120_REGS_EXT_PARTVERSION) are stored in that file. What I'm saying is we should also define the values for the data actually present at those addresses. For example, when we read from the register with the following 8-bit address: CC1120_REGS_EXT_PARTNUMBER, the data we get back should be 0x48 (An 8-bit number that depends on the transceiver. Since we're using the CC1120, the user's guide says the value is 0x48). 

Also, since the values I'm talking about aren't addresses, they shouldn't be stored in the regs file. It'd make more sense to store them in the cc1120_spi.h file.

**Comment by manumanuk** [comment]
# Current Status

Merging this PR because it is growing too large. I've decided that we can deal with issues later if we write them down now

## Purpose
- Create initial project setup for driver testing, including register definitions and defaults
- Write SPI interfacing modes and E2E tests for each type
- Create a logging function for easy logging
- Create abstraction for MCU functions to easily switch over to RM46

## New Changes
- Created project structure, including one folder for quick_tests just to check if the CC1120 can read data appropriately, and one folder for fully-fledged E2E testing & SPI functions
- Created SPI read, write, burst read/write, FIFO standard access, and FIFO direct access functions as specified in datasheet
- Created unit tests for each SPI function that can be run on the CC1120 to check if they are written correctly
- MCU abstraction functions and setup
- Logging functions
- Documentation + documentation standard

## Outstanding Changes
- Additional checks for SPI functions (e.g. if writing data larger than X, reject FIFO write)
- Improve unit tests
- Add enums for CC1120 state variable
- Additional documentation
- Add enums for error codes from SPI functions
- Fix logging vs mcu file structure - either separate the logging functions into the logging files, or combine the logging file into the mcu file
